vpc/service/internal_apis.go

```diff
@@ -407,3 +409,231 @@ func (vpcService *vpcService) doDetachBranchNetworkInterface(ctx context.Context
+	}
+
+	err = resetSgTx.Commit()
+	if err != nil {
+		err = errors.Wrap(err, "Could not commit transaction")
+		tracehelpers.SetStatus(err, span)
+		return nil, err
+	}
+
+	// Create the ec2 session to call AWS with the changed SG and before beginning the next Tx
+	resetSgSession, err := vpcService.ec2.GetSessionFromAccountAndRegion(ctx, ec2wrapper.Key{Region: region, AccountID: account})
+	if err != nil {
+		err = fmt.Errorf("Unable to create an ec2 session to reset SG %s : %w ", sgToDelete, err)
+		tracehelpers.SetStatus(err, span)
+		return &vpcapi.ResetSecurityGroupResponse{}, err
+	}
+
+	//Start new transaction
+	resetSgTx, err = beginSerializableTx(ctx, vpcService.db)
+	if err != nil {
+		err = errors.Wrap(err, "Unable to begin serializable transaction")
+		tracehelpers.SetStatus(err, span)
+		return &vpcapi.ResetSecurityGroupResponse{}, err
+	}
+
+	logger.G(ctx).WithField("resetSg", sgToDelete).Debug("Check if the SG to be reset is associated with a container")
+	// Check if the SG to be reset is associated with a container
+	rows, err := resetSgTx.QueryContext(ctx, `
+SELECT COUNT(*) FROM branch_enis
+WHERE ARRAY[$1] <@ security_groups AND branch_eni IN
+(SELECT branch_eni FROM branch_eni_attachments WHERE branch_eni_attachments.association_id IN
+	(SELECT branch_eni_association FROM assignments))
+		`, sgToDelete)
+	if err != nil {
+		err = fmt.Errorf("Could not query database for branch ENIs containing the SG %s: %w ", sgToDelete, err)
+		tracehelpers.SetStatus(err, span)
+		return &vpcapi.ResetSecurityGroupResponse{}, err
+	}
+	for rows.Next() {
+		err = rows.Scan(&count)
+		if err != nil {
+			err = fmt.Errorf("Could not get count of associations from query results %s: %w ", sgToDelete, err)
+			tracehelpers.SetStatus(err, span)
+			return &vpcapi.ResetSecurityGroupResponse{}, status.Errorf(codes.InvalidArgument,
+				"Could not get count of associations from query results")
+		}
+	}
+
+	if count != 0 {
+		//We cannot process the delete SG request as there are containers actively using the ENI that uses this SG
+		return &vpcapi.ResetSecurityGroupResponse{}, status.Errorf(codes.FailedPrecondition,
+			"%s is attached to an ENI with active association", sgToDelete)
+	}
+
+	// Reset the SG Id to the default SG
+	logger.G(ctx).WithField("resetSg", sgToDelete).Debug("Reset the SG Id to default")
+	rows, err = resetSgTx.QueryContext(ctx,
+		"UPDATE branch_enis SET security_groups = ARRAY[$1],dirty_security_groups=true WHERE ARRAY[$2] <@ security_groups returning branch_eni",
+		defaultSecurityGroupID, sgToDelete)
+	if err != nil {
+		err = fmt.Errorf("Cannot mark security groups as dirty for %s : %w", sgToDelete, err)
+		tracehelpers.SetStatus(err, span)
+		return &vpcapi.ResetSecurityGroupResponse{}, err
+	}
+
+	var enisWithSgToUpdate []string
+	for rows.Next() {
+		var updatedEni string
+		err := rows.Scan(&updatedEni)
+		if err != nil {
+			err = fmt.Errorf("Could not get VPC ID of %s: %w ", sgToDelete, err)
```

Comment:
I don't think this error string is correct.

Reply:
Changed.